cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -290,7 +290,27 @@ where
 				Ok(Some(local_receipt)) =>
 					return tx.send(Ok(local_receipt)).map_err(|_| GossipMessageError::SendError),
 				Ok(None) => {
-					tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+					// The local client has moved to the next block, that means the receipt
+					// of `block_hash` received from the network does not match the local one,
+					// we should just send back the local receipt at the same height.
```

Comment:
Hm, but there could be a legitimate fork, in which case this will cause ER to not be verified at all (rejected)?

Reply:
If an ER was from a legitimate fork, it'll still be verified but will be detected as invalid due to the verification failure. Currently, we process the ER on all the tips of primary chain, which should be revisited later I believe. https://github.com/subspace/subspace/blob/c428ffb11312078dd9028aa97b5f6384cf54c8d1/polkadot/node/collation-generation/src/lib.rs#L235-L238

Reply:
Hm, I think my question is different. Below you have a check `self.client.info().best_number >= block_number`, but there could be another temporary fork that is below best chain. Because of that check we will not wait for corresponding local receipt to be produced unless I misunderstood it.

Reply:
I don't quite understand which kind of fork you're referring to, but we indeed haven't thought about the edge cases.

Reply:
I'm talking about fork on the primary chain.

Reply:
A _corresponsing_ local receipt is always returned:

- If the ER for a given block hash is found, great, that means we have the same result locally.
- If the ER for a given block hash is not found, and we observe that the local client has progressed to the next block, that means we have different a different execution result, we return the local receipt at the same height. In this case, you don't have to wait as the local receipt had already been produced unless you're referring to another ER which is not this height.

Reply:
What if:
1. the primary chain has moved on to the next block
2. ER for a different fork for a previous block is received and discarded due to `if` statement
3. later reorg happens and chain needs to switch to new best chain

As the result the ER received in 2. will not be checked whether it was correct or not.

Reply:
Okay, I see your concern, but I don't have a clear answer right now. Theoretically, if the primary chain reorgs, the executor will reorg accordingly and the new receipt will be published to the network, so as long as the local executor reorgs in time, the example ER received in 2 will be checked again. We'll test these cases once the test framework is done(should be soon).

Reply:
Let's have a TODO above that `if` statement there, it seriously bothers me.

Reply:
TODO added.